app/components/request/config-form.js

```diff
@@ -0,0 +1,69 @@
+import Component from '@ember/component';
+import { computed } from '@ember/object';
+import { task } from 'ember-concurrency';
+import { reads } from '@ember/object/computed';
+import BranchSearching from 'travis/mixins/branch-searching';
+import { bindKeyboardShortcuts, unbindKeyboardShortcuts } from 'ember-keyboard-shortcuts';
+
+const CONFIG = {
+  SOURCE: {
+    API: 'api'
+  },
+  JAVASCRIPT: 'javascript',
+  YAML: 'yaml',
+  JSON: 'json',
+};
+
+export default Component.extend(BranchSearching, {
+  classNames: ['request-configs-form'],
+
+  repo: reads('request.repo'),
+
+  keyboardShortcuts: {
+    'shift+enter': 'submit'
+  },
+
+  onChange() {},
+  onSubmit() {},
+
+  didInsertElement() {
+    this._super(...arguments);
+    bindKeyboardShortcuts(this);
+  },
+
+  willDestroyElement() {
+    this._super(...arguments);
+    unbindKeyboardShortcuts(this);
+  },
+
+  searchBranches: task(function* (query) {
+    const result = yield this.searchBranch.perform(this.get('repo.id'), query);
+    return result.mapBy('name');
+  }),
+
+  configMode: computed('config', function () {
+    const { config } = this;
+    const { JAVASCRIPT, YAML } = CONFIG;
+    return config && config.startsWith('{') ? JAVASCRIPT : YAML;
+  }),
+
+  configs: computed('request.uniqRawConfigs', 'config', function () {
+    const { SOURCE } = CONFIG;
+    let configs = this.get('request.uniqRawConfigs') || [];
+    configs = configs.reject(config => config.source === SOURCE.API);
+    if (this.config) {
+      configs.unshift({ config: this.config, source: SOURCE.API });
```

Comment:
It's usually better to avoid using methods that mutate data, because they can accidentally leak unexpected changes to outer context. For example, if `configs.reject` a few lines above wouldn't create a copy of the array, this change would mutate `request.uniqRawConfigs` property, which, I guess, would not be expected.

Something like this would do the same, but without mutation:

```suggestion
      configs = [{ config: this.config, source: SOURCE.API }, ...configs];
```

Reply:
done: https://github.com/travis-ci/travis-web/commit/4c287dff491425537ff83707eaeba37b4f0b977a